toolkit/tools/internal/safechroot/safechroot.go

```diff
@@ -441,50 +446,82 @@ func cleanupAllChroots() {
 	inChrootMutex.Lock()
 
 	// mount is only supported in regular pipeline
+	failedToUnmount := false
 	if buildpipeline.IsRegularBuild() {
 		// Cleanup chroots in LIFO order incase any are interdependent (e.g. nested safe chroots)
 		logger.Log.Info("Cleaning up all active chroots")
 		for i := len(activeChroots) - 1; i >= 0; i-- {
 			logger.Log.Infof("Cleaning up chroot (%s)", activeChroots[i].rootDir)
-			err := activeChroots[i].unmountAndRemove(leaveChrootOnDisk)
+			err := activeChroots[i].unmountAndRemove(leaveChrootOnDisk, true)
 			// Perform best effort cleanup: unmount as many chroots as possible,
 			// even if one fails.
 			if err != nil {
 				logger.Log.Errorf("Failed to unmount chroot (%s)", activeChroots[i].rootDir)
+				failedToUnmount = true
 			}
 		}
 	}
 
-	logger.Log.Info("Cleanup finished")
+	if failedToUnmount {
+		logger.Log.Fatalf("Failed to unmount a chroot, manual unmount required. See above errors for details on which mounts failed.")
+	} else {
+		logger.Log.Info("Cleanup finished")
+	}
 }
 
 // unmountAndRemove retries to unmount directories that were mounted into
 // the chroot until the unmounts succeed or too many failed attempts.
 // This is to avoid leaving folders like /dev mounted when the chroot folder is forcefully deleted in cleanup.
 // Iff all mounts were successfully unmounted, the chroot's root directory will be removed if requested.
-func (c *Chroot) unmountAndRemove(leaveOnDisk bool) (err error) {
+func (c *Chroot) unmountAndRemove(leaveOnDisk, isFinalAttempt bool) (err error) {
```

Comment:
Nit-pick: on the level of this function I'm not sure if the final/non-final distinction makes sense. Seeing what the function's job is, it seems like lazy/non-lazy might be more suitable. Whether someone decided to do a lazy unmount as part of a final unmount attempt or not is outside what this function needs to worry about. What do you think?

Reply:
Sure, I can do that.